### ISSUE TITLE
Type not restricted after an equality check

### ISSUE THREAD

**Issue by angelo-mollame-2** [microsoft/TypeScript#36891]
**TypeScript Version:**  3.7.5

**Search Terms:** type not restricted after check

**Code**

```ts
function getFoo1(value: "A" | "B"): void {
    if (value === "A") {
        // This works
        const a: "A" = value;
    }
}
 
function getFoo2<T extends "A" | "B">(value: T): void {
    if (value === "A") {
        // But for some reason this doesn't.
        // Error: type T is not assignable to type "A"
        const a: "A" = value;
    }
}
```

**Expected behavior:** I should see no error

**Actual behavior:** error: type T is not assignable to type "A"

**Playground Link:** https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=14&pc=2#code/GYVwdgxgLglg9mABAcwKZQGJzgRgBQBuAhgDYioBciARAILWIA+NAQtQJRUFwwAmiAbwBQiUYhjBEhUuUQBeBTXrtBIseoD0GxABUAFjADOiAO5wATgGtDa9aIgJDURESp0GcxMTKoA3LcQAXyFgxCFQSFgEFHQsOAAmAB4dRFQADyhUMF5jdyZWagA+aR8qHU4vHn5hdQkpb1kFT3cVGrtRLUQWEGdgC0RDOABbVERzVCJBpCgDY144VEMwAHIoADoAzW0AUXNzCyooAE8AB1GUo0QwOGdJwxhkMCIAIxJRqDhEY7Olak2xBxgJwuNz0eReGR+ALBQJAA

**Comment by RyanCavanaugh** [comment]
Duplicate #32715 & others
